### PR TITLE
Use try-with-resources and canonical getters in TupleReader

### PR DESCRIPTION
**Pull request by srinicodebytes**
- Check for null
- Use try-with-resources
- Use right JDBC getXXX methods

### CODE REVIEW

File: src/com/google/enterprise/adaptor/database/TupleReader.java [hunk 1]
```diff
@@ -205,45 +289,53 @@ protected void emitRow(ResultSet resultSet, ContentHandler handler)
...
-          reader.close();
+          if (string != null) {
+            try (Writer writer =
+                new BufferedWriter(new ContentHandlerWriter(handler))) {
+              handler.startElement("", columnName, columnName, atts);
+              writer.write(string);
+            }
+          }
           break;
         case Types.NCHAR:
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/database/TupleReader.java]: Maybe we should be calling getNString on these first two, but we're using getString elsewhere, and I think we should do the same here (especially since we changed CHAR/VARCHAR to avoid using a stream).

File: src/com/google/enterprise/adaptor/database/TupleReader.java [hunk 2]
```diff
@@ -205,45 +289,53 @@ protected void emitRow(ResultSet resultSet, ContentHandler handler)
...
+              writer.write(String.valueOf(value));
+            }
+          }
           break;
         case Types.OTHER:
         default:
           string = resultSet.getString(col);
           if (null == string) {
             string = "" + resultSet.getObject(col);
+          } else {
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/database/TupleReader.java]: What? I think calling getObject is silly here, but you're changing the behavior of null here, too, but with strange code. Let's figure out the null handling, and then fix this.

File: src/com/google/enterprise/adaptor/database/TupleReader.java [hunk 3]
```diff
@@ -151,49 +154,130 @@ protected void emitRow(ResultSet resultSet, ContentHandler handler)
       if (resultSet.getObject(col) == null) {
         atts.addAttribute("", "ISNULL", "ISNULL", "NMTOKEN", "true");
         handler.startElement("", columnName, columnName, atts);
-        // TODO: Re-examine this need for flush.
-        outWriter.flush();
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/database/TupleReader.java]: Why did you remove the calls to flush?

File: src/com/google/enterprise/adaptor/database/TupleReader.java [hunk 4]
```diff
@@ -151,49 +154,130 @@ protected void emitRow(ResultSet resultSet, ContentHandler handler)
...
         handler.endElement("", columnName, columnName);
         continue;
       }
       switch (sqlType) {
         case Types.DATE:
           Date dateValue = resultSet.getDate(col);
-          handler.startElement("", columnName, columnName, atts);
-          outWriter.write(DATEFMT.format(dateValue));
+          if (dateValue != null) {
+            try (Writer writer =
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/database/TupleReader.java]: s/writer/outWriter/ (not enough reason to change even more lines beyond whitespace, which can be ignored by git blame).

File: src/com/google/enterprise/adaptor/database/TupleReader.java [hunk 5]
```diff
@@ -151,49 +154,130 @@ protected void emitRow(ResultSet resultSet, ContentHandler handler)
...
-        outWriter.flush();
         handler.endElement("", columnName, columnName);
         continue;
       }
       switch (sqlType) {
         case Types.DATE:
           Date dateValue = resultSet.getDate(col);
-          handler.startElement("", columnName, columnName, atts);
-          outWriter.write(DATEFMT.format(dateValue));
+          if (dateValue != null) {
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/database/TupleReader.java:158]: So you're changing the null handling here. Are there any tests of it? What does the change look like in the output? Are you delaying the construction of the writer just because the null handling is more localized?

File: src/com/google/enterprise/adaptor/database/TupleReader.java [hunk 6]
```diff
@@ -205,45 +289,53 @@ protected void emitRow(ResultSet resultSet, ContentHandler handler)
...
-          Reader nReader = resultSet.getNCharacterStream(col);
-          handler.startElement("", columnName, columnName, atts);
-          copyValidXMLCharacters(nReader, outWriter);
-          nReader.close();
+          try (Reader reader = resultSet.getNCharacterStream(col);
+              Writer writer =
+                  new BufferedWriter(new ContentHandlerWriter(handler))) {
+            if (reader != null) {
+              handler.startElement("", columnName, columnName, atts);
+              copyValidXMLCharacters(reader, writer);
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/database/TupleReader.java]: Another weirdness here, and maybe this is related to Lavanya's fix, is that we were using this copyValidXMLCharacters for streams, but we're not using it for strings. I think we should split out the *CHAR types, and use getString, but wrap them in a StringReader and call copyValidXMLCharacters. (We don't need it for the integer types, which can't have invalid XML characters to begin with.)

**Review by wiarlawd — CHANGES_REQUESTED** [review]
Mostly concerned about nulls, and the corresponding lack of test changes, indicating we probably need more tests.

File: src/com/google/enterprise/adaptor/database/TupleReader.java [hunk 7]
```diff
@@ -148,52 +152,127 @@ protected void emitRow(ResultSet resultSet, ContentHandler handler)
...
+            if (lob != null) {
+              // so encode using Base64
+              atts.addAttribute("", "encoding", "encoding", "NMTOKEN",
+                  "base64binary");
+              handler.startElement("", columnName, columnName, atts);
+              encode(lob, outWriter);
+            }
+          }
+          break;
+        case Types.LONGVARCHAR:
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/database/TupleReader.java]: Put this back down where it was, after VARCHAR.

File: src/com/google/enterprise/adaptor/database/TupleReader.java [hunk 8]
```diff
@@ -202,48 +281,68 @@ protected void emitRow(ResultSet resultSet, ContentHandler handler)
...
+              handler.startElement("", columnName, columnName, atts);
+              outWriter.write(string);
+            }
+          }
           break;
+        case Types.CHAR:
+        case Types.VARCHAR:
         case Types.NCHAR:
-        case Types.NCLOB:
+          String str = resultSet.getString(col);
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/database/TupleReader.java]: s/String str/string/, and the cascading diffs below.

File: src/com/google/enterprise/adaptor/database/TupleReader.java [hunk 9]
```diff
@@ -202,48 +281,68 @@ protected void emitRow(ResultSet resultSet, ContentHandler handler)
...
+            }
+          }
+          break;
         case Types.NVARCHAR:
         case Types.LONGNVARCHAR:
-          Reader nReader = resultSet.getNCharacterStream(col);
-          handler.startElement("", columnName, columnName, atts);
-          copyValidXMLCharacters(nReader, outWriter);
-          nReader.close();
+          try (Reader reader = resultSet.getNCharacterStream(col);
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/database/TupleReader.java]: s/reader/nReader/ here, but not in NCLOB.

File: src/com/google/enterprise/adaptor/database/TupleReader.java [hunk 10]
```diff
@@ -202,48 +281,68 @@ protected void emitRow(ResultSet resultSet, ContentHandler handler)
...
+          String str = resultSet.getString(col);
+          if (str != null) {
+            try (Writer outWriter =
+                new BufferedWriter(new ContentHandlerWriter(handler))) {
+              handler.startElement("", columnName, columnName, atts);
+              copyValidXMLCharacters(new StringReader(str), outWriter);
+            }
+          }
+          break;
         case Types.NVARCHAR:
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/database/TupleReader.java]: Move this to the previous case.

File: src/com/google/enterprise/adaptor/database/TupleReader.java [hunk 11]
```diff
@@ -202,48 +281,68 @@ protected void emitRow(ResultSet resultSet, ContentHandler handler)
...
+              outWriter.write(String.valueOf(value));
+            }
+          }
           break;
         case Types.OTHER:
         default:
-          string = resultSet.getString(col);
-          if (null == string) {
-            string = "" + resultSet.getObject(col);
+          str = resultSet.getString(col);
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/database/TupleReader.java]: s/str/string/ throughout this case.

File: src/com/google/enterprise/adaptor/database/TupleReader.java (hunk 5 above)
Comment (BrettMichaelJohnson) [reply src/com/google/enterprise/adaptor/database/TupleReader.java:158]: Also related to null checking, the check sqlTypeName != null 3 lines up is redundant.  It cannot be null here. (FindBugs)

File: test/com/google/enterprise/adaptor/database/TupleReaderTest.java [hunk 12]
```diff
@@ -118,6 +118,24 @@ public void testVARCHAR() throws Exception {
     assertEquals(golden, result);
   }
 
+  @Test
+  public void testNullVARCHAR() throws Exception {
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/TupleReaderTest.java]: testVarchar_null, here and throughout, but don't change the existing test names.

Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/TupleReaderTest.java]: This test case already existed (as testNULL), so you might as well delete testNULL.

**Review by BrettMichaelJohnson — COMMENTED** [review]
Other issues flagged by FindBugs:

TupleReader redundantly implements XMLReader (already implemented by XMLFIlterImpl).

The SimpleDateFormats (DATEFORMAT, TIMEFORMAT, TIMEZONEFORMAT) are not thread-safe.

File: test/com/google/enterprise/adaptor/database/TupleReaderTest.java [hunk 13]
```diff
@@ -118,6 +118,24 @@ public void testVARCHAR() throws Exception {
...
+    executeUpdate("create table data(id integer, colname varchar)");
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/TupleReaderTest.java]: What's your reasoning for wanting the ID columns? Having a test with multiple columns seems good, as does a test with mixed null/non-null values, but it doesn't need to be replicated for each type. I would drop the ID columns in these null tests and add a single test with a few columns of duplicate and different types, with some nulls mixed in.

File: src/com/google/enterprise/adaptor/database/TupleReader.java [hunk 14]
```diff
@@ -148,52 +152,127 @@ protected void emitRow(ResultSet resultSet, ContentHandler handler)
...
+            } finally {
+              try {
+                nclob.free();
+              } catch (Exception e) {
+                log.log(Level.FINEST, "Error closing NCLOB", e);
+              }
+            }
+          }
+          break;
         case Types.VARBINARY:
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/database/TupleReader.java]: Add BINARY.

**Review by wiarlawd — CHANGES_REQUESTED** [review]
Also, add test pairs (non-null/null) for CLOB, BINARY, and BOOLEAN.

**Comment by wiarlawd** [comment]
FYI, I do not want to fix the FindBugs errors in this CL. Two of them are not very interesting, and the thread-safety deserves its own CL.

File: test/com/google/enterprise/adaptor/database/TupleReaderTest.java [hunk 15]
```diff
@@ -213,6 +190,43 @@ public void testNullINTEGER() throws Exception {
   }
 
   /**
+   * Test boolean column.
+   */
+  @Test
+  public void testBOOLEAN() throws Exception {
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/TupleReaderTest.java]: testBoolean

File: test/com/google/enterprise/adaptor/database/TupleReaderTest.java [hunk 16]
```diff
@@ -213,6 +190,43 @@ public void testNullINTEGER() throws Exception {
...
+        + "</table_rec>"
+        + "</table>"
+        + "</database>";
+    ResultSet rs = executeQueryAndNext("select * from data");
+    String result = generateXml(rs);
+    assertEquals(golden, result);
+  }
+
+  @Test
+  public void testNullBOOLEAN() throws Exception {
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/TupleReaderTest.java]: testBoolean_null

File: test/com/google/enterprise/adaptor/database/TupleReaderTest.java [hunk 17]
```diff
@@ -158,13 +137,12 @@ public void testChar() throws Exception {
 
   @Test
   public void testNullChar() throws Exception {
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/TupleReaderTest.java]: testChar_null, that's what I meant by "throughout"

File: test/com/google/enterprise/adaptor/database/TupleReaderTest.java [hunk 18]
```diff
@@ -428,4 +438,86 @@ public void testNullBLOB() throws Exception {
     String result = generateXml(rs);
     assertEquals(golden, result);
   }
+
+  /**
+   * Test CLOB column.
+   */
+  @Test
+  public void testCLOB() throws Exception {
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/TupleReaderTest.java]: testClob

File: test/com/google/enterprise/adaptor/database/TupleReaderTest.java [hunk 19]
```diff
@@ -428,4 +438,86 @@ public void testNullBLOB() throws Exception {
...
+            + " would consider on a case-by-case basis requests to"
+            + " remove the link to that site from its indices. However,"
+            + " if the operator of the site does not take steps to"
+            + " prevent it, the automatic facilities used to create"
+            + " the indices are likely to find that site and index it"
+            + " again in a relatively short amount of time.";
+    executeUpdate("create table data(colname clob)");
+    String sql = "insert into data(colname) values (?)";
+    try (PreparedStatement ps = getConnection().prepareStatement(sql)) {
+      ps.setBinaryStream(1, new ByteArrayInputStream(clobData.getBytes(UTF_8)));
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/TupleReaderTest.java]: setString(1, clobData) is fine, but never setBinaryStream.

File: test/com/google/enterprise/adaptor/database/TupleReaderTest.java [hunk 20]
```diff
@@ -428,4 +438,86 @@ public void testNullBLOB() throws Exception {
...
+        + "</table_rec>"
+        + "</table>"
+        + "</database>";
+    ResultSet rs = executeQueryAndNext("select * from data");
+    String result = generateXml(rs);
+    assertEquals(golden, result);
+  }
+
+  @Test
+  public void testmultipleTypes() throws Exception {
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/TupleReaderTest.java]: Capitalize "M"

**Review by wiarlawd — CHANGES_REQUESTED** [review]
One coding issue, the rest are names. Please changed added tests everywhere to match the style guide ("testBOOLEAN" is not allowed) and our test conventions ("testSomething_testCase" rather than "testTestCaseSomething").

File: src/com/google/enterprise/adaptor/database/TupleReader.java [hunk 21]
```diff
@@ -202,48 +272,78 @@ protected void emitRow(ResultSet resultSet, ContentHandler handler)
...
+          }
           break;
-        case Types.CLOB:
         case Types.LONGVARCHAR:
-          Reader reader = resultSet.getCharacterStream(col);
-          handler.startElement("", columnName, columnName, atts);
-          copyValidXMLCharacters(reader, outWriter);
-          reader.close();
+          try (Reader reader = resultSet.getCharacterStream(col);
+              Writer writer =
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/database/TupleReader.java]: s/writer/outWriter/ for this case.

File: test/com/google/enterprise/adaptor/database/TupleReaderTest.java [hunk 22]
```diff
@@ -287,4 +421,103 @@ public void testBLOB() throws Exception {
...
+        + "</table>"
+        + "</database>";
+    ResultSet rs = executeQueryAndNext("select * from data");
+    String result = generateXml(rs);
+    assertEquals(golden, result);
+  }
+
+  @Test
+  public void testMultipleTypes() throws Exception {
+    executeUpdate("create table data(id integer, name varchar, "
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/TupleReaderTest.java]: Break before these two string to avoid string concatenation (prefer higher-level breaks).

**Review by wiarlawd — APPROVED** [review]
LGTM++ with two nits.

Please change commit title to "Use try-with-resources and canonical getters in TupleReader" and remove the description entirely. This is more specific about the changes.
